velox/functions/sparksql/String.h

```diff
@@ -1431,4 +1432,211 @@ struct LevenshteinDistanceFunction {
+      const arg_type<Varchar>* upperCharPtr,
+      const arg_type<Varchar>* lowerCharPtr,
+      const arg_type<Varchar>* digitCharPtr,
+      const arg_type<Varchar>* otherCharPtr) {
+    if (inputPtr == nullptr) {
+      return false;
+    }
+
+    doCall(
+        result,
+        *inputPtr,
+        getMaskedChar(upperCharPtr),
+        getMaskedChar(lowerCharPtr),
+        getMaskedChar(digitCharPtr),
+        getMaskedChar(otherCharPtr));
+    return true;
+  }
+
+ private:
+  void doCall(
+      out_type<Varchar>& result,
+      StringView input,
+      const std::optional<StringView> upperChar,
+      const std::optional<StringView> lowerChar,
+      const std::optional<StringView> digitChar,
+      const std::optional<StringView> otherChar) const {
+    auto inputBuffer = input.data();
+    const size_t inputSize = input.size();
+    result.reserve(inputSize);
+    auto outputBuffer = result.data();
+    size_t inputIdx = 0;
+    size_t outputIdx = 0;
+    while (inputIdx < inputSize) {
+      utf8proc_int32_t curCodePoint;
+      int charByteSize;
+      curCodePoint = utf8proc_codepoint(
+          &inputBuffer[inputIdx], inputBuffer + inputSize, charByteSize);
+      if (curCodePoint == -1) {
+        // That means it is a invalid UTF-8 character for example '\xED',
+        // treat it as char with size is 1.
+        charByteSize = 1;
+      }
+      auto maskedChar = &inputBuffer[inputIdx];
+      auto maskedCharByteSize = charByteSize;
+      // Treat invalid UTF-8 character as other char.
+      utf8proc_propval_t category = utf8proc_category(curCodePoint);
+      if (isUpperChar(category) && upperChar.has_value()) {
+        maskedChar = upperChar.value().data();
+        maskedCharByteSize = upperChar.value().size();
+      } else if (isLowerChar(category) && lowerChar.has_value()) {
+        maskedChar = lowerChar.value().data();
+        maskedCharByteSize = lowerChar.value().size();
+      } else if (isDigitChar(category) && digitChar.has_value()) {
+        maskedChar = digitChar.value().data();
+        maskedCharByteSize = digitChar.value().size();
+      } else if (
+          !isUpperChar(category) && !isLowerChar(category) &&
+          !isDigitChar(category) && otherChar.has_value()) {
+        maskedChar = otherChar.value().data();
+        maskedCharByteSize = otherChar.value().size();
+      }
+
+      for (auto i = 0; i < maskedCharByteSize; i++) {
+        outputBuffer[outputIdx++] = maskedChar[i];
+      }
+
+      inputIdx += charByteSize;
+    }
+    result.resize(outputIdx);
+  }
+
+  bool isUpperChar(utf8proc_propval_t& category) const {
+    return category == UTF8PROC_CATEGORY_LU;
+  }
+
+  bool isLowerChar(utf8proc_propval_t& category) const {
+    return category == UTF8PROC_CATEGORY_LL;
+  }
+
+  bool isDigitChar(utf8proc_propval_t& category) const {
+    return category == UTF8PROC_CATEGORY_ND;
+  }
+
+  std::optional<StringView> getMaskedChar(const arg_type<Varchar>* maskChars) {
```

Comment:
nit: maskChars points to a single value, and shall we rename it as maskChar? Ditto for the other variables.